Add Pass to move constructors from cpu to cuda

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109665

Sometimes indexing tensors are constructed on cpu and then used to index a cuda tensor. This prevents cudagraphs when it does not need to. Adding a pass which moves constructors from cpu->cuda when we can prove the downstream uses can be safely converted.

This pr allows us to cudagraph `clip` from the blueberries model which improves perf from ~1.5x -> ~4x.




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel

eellison committed:
Add Pass to move constructors from cpu to cuda

[ghstack-poisoned]

eellison committed:
Update on "Add Pass to move constructors from cpu to cuda"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Pass to move constructors from cpu to cuda"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang commented:
I'm not too familiar with inductor's fx passes. Would you like to teach me how they work, or are there other bits you'd like me to review?

peterbell10 commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -802,3 +805,164 @@ def repl(inp, x1, x2):
+
+    for node in movable_constructors:
+        kwargs = node.kwargs.copy()
+        kwargs["device"] = next(iter(cuda_devices))
```
Shouldn't we track which device the tensor is copied to and construct it there?

eellison replied:
this pass doesn't handle multiple cuda devices so there should only ever be one right now

eellison commented:
@ezyang there shouldn't be too much inductor specific logic here - it could feasibly be a pass on the joint graph as well.

eellison committed:
Update on "Add Pass to move constructors from cpu to cuda"


Sometimes indexing tensors are constructed on cpu and then used to index a cuda tensor. This prevents cudagraphs when it does not need to. Adding a pass which moves constructors from cpu->cuda when we can prove the downstream uses can be safely converted.

This pr allows us to cudagraph `clip` from the blueberries model which improves perf from ~1.5x -> ~4x.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
+    )
+
+
+def cannot_be_moved_to_cuda(node):
+    if node.target == "output":
+        return True
+
+    if not isinstance(
+        node.target, torch._ops.OpOverload
+    ) and node.target.namespace not in ("prims", "aten"):
+        return True
```
cc @ezyang i'm restricting to aten/prims here but maybe i should be even more selective ? wdyt ?

ezyang replied:
I think the main question is whether or not to allowlist or blocklist at a fine-grained level. It seems... safer to allowlist on a per operator basis? I'm just thinking of situations where it's not OK to move an operator to CUDA and there's lots: (1) the operator is CPU only, (2) the CUDA version has drastically different numerics, (3) the CPU version is faster than the CUDA version / the CUDA version is crappy...

That being said, I think any operator that we have a lowering for is probably OK to move to CUDA. So you can get a decently sized list that way, probably.

ezyang replied:
It feels like we may have to collect some data on this

eellison replied:
only including operators inductor has lowerings for - i think that should work

ezyang commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
+        aten.index_put_.default,
+        aten.copy.default,
+        aten.copy_.default,
+        aten.slice_scatter.default,
```
make this a tag 🤔

eellison replied:
if we get another use of it sure 😛

ezyang commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
+    return None if not isinstance(ten, torch.Tensor) else ten.device
+
+
+def get_cpu_indeg_count(graph) -> Dict[fx.Node, int]:
```
in degree? (a brief comment here would help a lot!)

ezyang commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -86,6 +87,8 @@ def post_grad_passes(gm: torch.fx.GraphModule, is_inference: bool):
 
     stable_topological_sort(gm.graph)
 
+    move_constructors_to_cuda(gm.graph)
```
is fake tensor metadata up to date at this point? You need accurate device info right?

eellison replied:
we don't have any other device-changing optimizations so it should be..

ezyang commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
+
+
+def get_cpu_indeg_count(graph) -> Dict[fx.Node, int]:
+    cpu_indeg: Dict[fx.Node, int] = Counter()
```
You don't... actually use the Counter functionality do you??

ezyang commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
+            device = get_node_device(node)
+            cpu_count += device is not None and device.type == "cpu"
+
+        pytree.tree_map_only(torch.fx.Node, add_cpu_inp, (node.args, node.kwargs))
```
nit: I think map_aggregate is good enough here

ezyang commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
+            unioned_set = make_dependencies_equivalent(
+                dependencies, constructor_dependencies[user]
+            )
+            constructor_dependencies[user] = unioned_set
```
it's weird to me that you didn't fold this into the function

eellison replied:
kind of leaves it open for a more efficient impl (union find) if we needed it.. but yea fair

ezyang commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
+        # could use union find but not worth complexity here
+        set1.update(set2)
+        for obj in set1:
+            equal_constructor_sets[obj] = set1
```
As far as I can tell, you never actually read from this data structure. What's up with that?

eellison replied:
Good catch - fixed

ezyang commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
+def find_movable_constructors(graph, constructors: List[fx.Node]) -> Set[fx.Node]:
+    """
+    Starting from the cpu constructors, iterate through the graph and test that all of their
+    downstream uses can safely be moved to cpu.
```
Maybe this is just a matter of taste, but to me it seems like it would make more sense to do this as a backwards pass: e.g., we start with the cpu-cuda transition barrier, and then we try to successively absorb cpu operations into the cuda region, so long as doing so wouldn't increase DtoH memory traffic.

eellison replied:
This is a good suggestion - for now i'm going to see how dashboard looks for current impl. Some of the other backends (xla) may want to be more aggressive and not include that.

ezyang commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
+        node.kwargs = kwargs
+
+
+def find_movable_constructors(graph, constructors: List[fx.Node]) -> Set[fx.Node]:
```
constructors is required to be in order, is that correct?

eellison replied:
shouldn't actually matter here

ezyang commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
+
+    # if a cpu node has a dependency on two different cpu constructors,
+    # then if either constructor cannot be moved to cuda, the other cannot as well.
+    # In this case any node with a dependency on one will have a dependency on the other
```
Unless you're willing to duplicate constructor calls, no?

eellison replied:
While that is true, we're mostly optimizing for cudagraphs so going to leave this unimplemented for now

ezyang reviewed (approved):
Seems worth trying

eellison commented:
@pytorchbot rebase

eellison commented:
@pytorchbot rebase

eellison committed:
Update on "Add Pass to move constructors from cpu to cuda"


Sometimes indexing tensors are constructed on cpu and then used to index a cuda tensor. This prevents cudagraphs when it does not need to. Adding a pass which moves constructors from cpu->cuda when we can prove the downstream uses can be safely converted.

This pr allows us to cudagraph `clip` from the blueberries model which improves perf from ~1.5x -> ~4x.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Pass to move constructors from cpu to cuda"


Sometimes indexing tensors are constructed on cpu and then used to index a cuda tensor. This prevents cudagraphs when it does not need to. Adding a pass which moves constructors from cpu->cuda when we can prove the downstream uses can be safely converted.

This pr allows us to cudagraph `clip` from the blueberries model which improves perf from ~1.5x -> ~4x.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Pass to move constructors from cpu to cuda"


Sometimes indexing tensors are constructed on cpu and then used to index a cuda tensor. This prevents cudagraphs when it does not need to. Adding a pass which moves constructors from cpu->cuda when we can prove the downstream uses can be safely converted.

This pr allows us to cudagraph `clip` from the blueberries model which improves perf from ~1.5x -> ~4x.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented:
Improves sam 14%, landing !

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot revert -m "want to make one last change before landing"

eellison commented:
@pytorchbot revert -m "want to maek one last change" -c weird

eellison committed:
Update on "Add Pass to move constructors from cpu to cuda"


Sometimes indexing tensors are constructed on cpu and then used to index a cuda tensor. This prevents cudagraphs when it does not need to. Adding a pass which moves constructors from cpu->cuda when we can prove the downstream uses can be safely converted.

This pr allows us to cudagraph `clip` from the blueberries model which improves perf from ~1.5x -> ~4x.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler Xia-Weiwen ngimel

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
